### ISSUE TITLE
[libpqxx] pqxx::connection fails with double free or corruption (out) after upgrade from 7.7.1 to 7.8.2

### ISSUE THREAD

**Issue by chr1stobald** [microsoft/vcpkg#36168]
**Describe the bug**

**Environment**
- OS: Linux Ubuntu LTS 22.04.3 (WSL)
- Compiler: GNU 11 or CLANG 14 (C++17 or 20)

**To Reproduce**
1. ./vcpkg install libpqxx

#include <iostream>
#include <string>
#include<pqxx/pqxx>

int main()
{
      std::string cnxUri = std::getenv("PG_CNX_URI");
      pqxx::connection cnx(cnxUri);
}

// cnxUri like postgresql://user:password@127.0.0.1/database

3. See error

double free or corruption (out)
Aborted

Repro code when

**Expected behavior**

No crash

**Failure logs**
Failed in fe-connect.c->freePGconn(PGConn *conn)

4413 free(conn->pghost)

conn is not nullptr but pghost contains garbage. 

**Additional context**
PostgreSQL 14.
After git pull then vcpkg upgrade then complete re-install from scratch of vcpkg

**Comment by chr1stobald** [comment]
Hello Monica,

Thanks for the prompt response. Still I believe there's a problem maybe
specific tp vcpkg running on Ubuntu 22.04.3 LTS via WSL2.
***@***.***:~/workspace/testpqxx$ cat /etc/*-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04.3 LTS"
PRETTY_NAME="Ubuntu 22.04.3 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.3 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="
https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy

Try the following please:


*CMakeLists.txt*

cmake_minimum_required (VERSION 3.8)

set(CMAKE_TOOLCHAIN_FILE *~*/vcpkg/scripts/buildsystems/vcpkg.cmake CACHE
STRING "Vcpkg toolchain file")
#
project (test)
# Set the C++ standard flag
set (CMAKE_CXX_STANDARD 17)
set (CMAKE_CXX_STANDARD_REQUIRED YES)
set (CMAKE_CXX_EXTENSIONS NO)

####################################
#          CONFIG HERE             #
# Build a Debug or Release version #
####################################
set (CMAKE_BUILD_TYPE Debug)

find_package(libpqxx CONFIG REQUIRED)

# Adding Source files and libraries
add_executable(test test.cpp)
# Set compiler definition. Typical
add_definitions("-pedantic -W -Wall -Wextra -Wold-style-cast -march=native")
target_link_libraries(test PRIVATE libpqxx::pqxx)

if  (CMAKE_BUILD_TYPE STREQUAL Debug)
    message("+-----------------------------------+")
    message("| BUILDING => test Debug Version |")
    message("+-----------------------------------+")
    set(CMAKE_CXX_FLAGS "-O0 -ggdb3")
endif ()

if (CMAKE_BUILD_TYPE STREQUAL Release)
    message("+-------------------------------------+")
    message("| BUILDING => test Release Version |")
    message("+-------------------------------------+")
    set(CMAKE_CXX_FLAGS "-O3 -flto")
endif ()

*test.cpp*

Make sure PG_YOUR_INFO is valid.

#include <iostream>
#include <pqxx/pqxx>

int main()
{
    std::string cnxUri = std::getenv("PG_YOUR_INFO");
    pqxx::connection cnx(cnxUri);

    return 0;
}

~/workspace/testpqxx/build$ ./test
double free or corruption (out)
Aborted

Verified, same result with CLANG or G++.

Thanks for the help,

Regards,

Christophe Fraule

**Comment by MonicaLiu0311** [comment]
The test also passes:
```
monica@monica:/mnt/vcpkg/test/build$ cmake ..
-- The C compiler identification is GNU 11.4.0
-- The CXX compiler identification is GNU 11.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PostgreSQL: /mnt/vcpkg/installed/x64-linux/debug/lib/libpq.a (found version "16.0")
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Found OpenSSL: /mnt/vcpkg/installed/x64-linux/debug/lib/libcrypto.a (found version "3.2.0")
+-----------------------------------+
| BUILDING => test Debug Version |
+-----------------------------------+
-- Configuring done
-- Generating done
-- Build files have been written to: /mnt/vcpkg/test/build
monica@monica:/mnt/vcpkg/test/build$
monica@monica:/mnt/vcpkg/test/build$
monica@monica:/mnt/vcpkg/test/build$
monica@monica:/mnt/vcpkg/test/build$ cmake --build .
[ 50%] Building CXX object CMakeFiles/test.dir/test.cpp.o
[100%] Linking CXX executable test
[100%] Built target test
```

vcpkg version:
```
monica@monica:/mnt/vcpkg$ ./vcpkg version
vcpkg package management program version 2024-01-11-710a3116bbd615864eef5f9010af178034cb9b44

See LICENSE.txt for license information.
```
environment:
```
monica@monica:/mnt/vcpkg$
monica@monica:/mnt/vcpkg$ cat /etc/*-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04.3 LTS"
PRETTY_NAME="Ubuntu 22.04.3 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.3 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```

**Comment by chr1stobald** [comment]
Hello Monica,

Did you test with vcpkg installed on WSL2/Ubuntu 22.04 LTS?

On my WSL2/Ubuntu VM /mnt points to the windows C: and WSL drive.

Thanks for the help and patience.

Regards,

Christophe

**Comment by MonicaLiu0311** [comment]
My machine is Ubuntu 22.04, but not WSL2.

**Comment by chr1stobald** [comment]
Hello Monica,

Could you make an attempt on a Ubuntu machine running 22.04 on WSL2? Vith
both vcpkg and the postgreSQL instances running in the same environment.

Thanks,

Christophe

**Comment by chr1stobald** [comment]
Hello Monica,

Were you able to run the same test with Ubuntu 22.04 on WSL2 (all on WSL2, vcpkg,  program and database).

Thanks for sharing an update,

Regards,

Christophe


Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by MonicaLiu0311** [comment]
@chr1stobald 

I finally tried it again today using WSL, but it worked just as well:
```
fx@fx001:~/Monica/vcpkg/test/build$ cmake ..
-- The C compiler identification is GNU 11.4.0
-- The CXX compiler identification is GNU 11.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PostgreSQL: /home/fx/Monica/vcpkg/installed/x64-linux/debug/lib/libpq.a (found version "16.0")
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Found OpenSSL: /home/fx/Monica/vcpkg/installed/x64-linux/debug/lib/libcrypto.a (found version "3.2.0")
+-----------------------------------+
| BUILDING => test Debug Version |
+-----------------------------------+
-- Configuring done
-- Generating done
-- Build files have been written to: /home/fx/Monica/vcpkg/test/build
fx@fx001:~/Monica/vcpkg/test/build$ cmake --build .
[ 50%] Building CXX object CMakeFiles/test.dir/test.cpp.o
[100%] Linking CXX executable test
[100%] Built target test
```

vcpkg version:
```
fx@fx001:~/Monica/vcpkg$ ./vcpkg version
vcpkg package management program version 2024-01-11-710a3116bbd615864eef5f9010af178034cb9b44

See LICENSE.txt for license information.
```
environment:
```
fx@fx001:~/Monica/vcpkg$ cat /etc/*-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04.3 LTS"
PRETTY_NAME="Ubuntu 22.04.3 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.3 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```

**Comment by SlateyDev** [comment]
I am getting this same error on 7.8.1 when trying to connect to a local db on Pop!_OS 22.04 LTS

**Comment by romanpunia** [comment]
Hello, guys,

I have encountered the same problem but with C version of this library - libpq. As i understand after OpenSSL updated their library to 3.2 libpq started to fail when working with some BIO related API. I found out that this was a known bug and can be fixed either by downgrading OpenSSL to 3.1 or older, or by upgrading liqpq to 16.2 or higher (which received a fix by PQ devs).

Because current libpq version is 16.0, there is only one solution at the moment - downgrade OpenSSL. As far as i know, code change may help only if dealing with blocking connections somehow and even that is random because there is a memory error involved.

Waiting on lipq 16.2 port.

Hope this helps.
